cli/cmd/check_extensions.go

```diff
@@ -0,0 +1,319 @@
+	cliExtensions := findCLIExtensionsOnPath(pathEnv, glob, exec)
+
+	// first, collect config extensions that are "always" enabled
+	extensions, checksSeen := findAlwaysChecks(cliExtensions, exec)
+
+	// nsLabelSet keeps track of the extension namespaces that we are still
+	// searching for an executable for.
+	nsLabelSet := map[string]struct{}{}
+	for _, label := range nsLabels {
+		if _, ok := checksSeen[label]; !ok {
+			nsLabelSet[label] = struct{}{}
+		}
+	}
+
+	// second, collect on-cluster extensions
+	for _, e := range cliExtensions {
+		suffix := suffix(e)
+		if _, ok := nsLabelSet[suffix]; ok {
+			extensions = append(extensions, extension{path: e})
+			delete(nsLabelSet, suffix)
+		}
+	}
+
+	// third, collect built-in extensions
+	for label := range nsLabelSet {
+		if _, ok := builtInChecks[label]; ok {
+			extensions = append(extensions, extension{path: os.Args[0], builtin: label})
+			delete(nsLabelSet, label)
+		}
+	}
+
+	// anything left in nsLabelSet is a missing executable
+	missing := []string{}
+	for label := range nsLabelSet {
+		missing = append(missing, fmt.Sprintf("linkerd-%s", label))
+	}
+
+	sort.Slice(extensions, func(i, j int) bool {
+		if extensions[i].path != extensions[j].path {
+			_, filename1 := filepath.Split(extensions[i].path)
+			_, filename2 := filepath.Split(extensions[j].path)
+			return filename1 < filename2
+		}
+		return extensions[i].builtin < extensions[j].builtin
+	})
+	sort.Strings(missing)
+
+	return extensions, missing
+}
+
+// findCLIExtensionsOnPath searches the path for all linkerd-* executables and
+// returns a slice of unique filepaths. if multiple executables have the same
+// name, only the one which comes earliest in the pathEnv is returned.
+func findCLIExtensionsOnPath(pathEnv string, glob glob, exec utilsexec.Interface) []string {
+	executables := []string{}
+	seen := map[string]struct{}{}
+
+	for _, dir := range filepath.SplitList(pathEnv) {
+		matches, err := glob(filepath.Join(dir, "linkerd-*"))
+		if err != nil {
+			continue
+		}
+		sort.Strings(matches)
+
+		for _, match := range matches {
+			suffix := suffix(match)
+			if _, ok := seen[suffix]; ok {
+				continue
+			}
+
+			path, err := exec.LookPath(match)
+			if err != nil {
+				continue
+			}
+
+			executables = append(executables, path)
+			seen[suffix] = struct{}{}
+		}
+	}
+
+	return executables
+}
+
+// findAlwaysChecks filters a slice of linkerd-* executables to only those that
+// support the "config" subcommand, and announce themselves to "always" run. the
+// checksSeen map returned informs the caller which extensions were identified
+// to always run, and therefore do not need to be evaluated for inclusion based
+// on on-cluster resources.
+func findAlwaysChecks(cliExtensions []string, exec utilsexec.Interface) ([]extension, map[string]struct{}) {
+	extensions := []extension{}
+
+	checksSeen := map[string]struct{}{}
+
+	for _, e := range cliExtensions {
+		suffix := suffix(e)
+		if _, ok := checksSeen[suffix]; ok {
```

Comment:
since findCLIExtensionsOnPath only returns one extension of each name, do we need this deduping logic?